packages/grid/x-data-grid/src/hooks/core/strategyProcessing/useGridStrategyProcessing.ts

```diff
@@ -0,0 +1,152 @@
+import { GridEvents } from '../../../models/events';
+
+export const GRID_DEFAULT_STRATEGY = 'none';
+
+export const GRID_STRATEGIES_PROCESSORS: {
+  [P in GridStrategyProcessorName]: GridStrategyProcessingLookup[P]['group'];
+} = {
+  rowTreeCreation: 'rowTree',
+  filtering: 'rowTree',
+  sorting: 'rowTree',
+};
+
+/**
+ * Implements a variant of the Strategy Pattern (see https://en.wikipedia.org/wiki/Strategy_pattern)
+ *
+ * Some plugins contain custom logic that must only be run if the right strategy is active.
+ * For instance, the tree data plugin has:
+ * - custom row tree creation algorithm.
+ * - custom sorting algorithm.
+ * - custom filtering algorithm.
+ *
+ * These plugins must use:
+ * - `apiRef.current.unstable_registerStrategyProcessor` to register their processors.
+ * - `apiRef.current.unstable_setStrategyAvailability` to tell if their strategy can be used.
+ *
+ * Some hooks need to run the custom logic of the active strategy.
+ * For instance, the `useGridFiltering` wants to run:
+ * - the tree data filtering if the tree data is the current way of grouping rows.
+ * - the row grouping filtering if the row grouping is the current way of grouping rows.
+ * - the flat filtering if there is no grouping of the rows (equivalent to the "none" strategy).
+ *
+ * These hooks must use:
+ * - `apiRef.current.unstable_applyStrategyProcessor` to run a processor.
+ * - `GridEvents.strategyActivityChange` to update something when the active strategy changes.
+ *    Warning: Be careful not to apply the processor several times.
+ *    For instance `GridEvents.rowsSet` is fired by `useGridRows` whenever the active strategy changes.
+ *    So listening to both would most likely run your logic twice.
+ * - `GridEvents.activeStrategyProcessorChange` to update something when the processor of the active strategy changes
+ *
+ * Each processor name is part of a strategy group which can only have one active strategy at the time.
+ */
+export const useGridStrategyProcessing = (apiRef: React.MutableRefObject<GridApiCommunity>) => {
+  const availableStrategies = React.useRef(
+    new Map<string, { group: GridStrategyGroup; isAvailable: () => boolean }>(),
+  );
+  const strategiesCache = React.useRef<{
+    [P in GridStrategyProcessorName]?: { [strategyName: string]: GridStrategyProcessor<P> };
+  }>({});
+
+  const registerStrategyProcessor = React.useCallback<
+    GridStrategyProcessingApi['unstable_registerStrategyProcessor']
+  >(
+    (strategyName, processorName, processor: GridStrategyProcessor<any>) => {
+      if (!strategiesCache.current[processorName]) {
+        strategiesCache.current[processorName] = {};
+      }
+
+      const groupPreProcessors = strategiesCache.current[processorName] as {
+        [strategyName: string]: GridStrategyProcessor<any>;
+      };
+      const previousProcessor = groupPreProcessors[strategyName];
+      if (previousProcessor !== processor) {
```

Comment:
You could early return the unmount function here if `previousProcessor === processor`. The same if `previousProcessor` is not defined.

Reply:
Not sure it is a lot cleaner

We still have to return the same cleanup function when the processing did not change.